// ng-connection-service/bundles/ng-connection-service.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('ng-connection-service', ['exports', '@angular/core', 'rxjs'], factory) :
    (factory((global['ng-connection-service'] = {}),global.ng.core,global.rxjs));
}(this, (function (exports,i0,rxjs) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ConnectionService = (function () {
        function ConnectionService() {
            this.connectionMonitor = new rxjs.Observable(function (observer) {
                window.addEventListener('offline', function (e) {
                    observer.next(false);
                });
                window.addEventListener('online', function (e) {
                    observer.next(true);
                });
            });
        }
        /**
         * @return {?}
         */
        ConnectionService.prototype.monitor = /**
         * @return {?}
         */
            function () {
                return this.connectionMonitor;
            };
        ConnectionService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] },
        ];
        /** @nocollapse */
        ConnectionService.ctorParameters = function () { return []; };
        /** @nocollapse */ ConnectionService.ngInjectableDef = i0.defineInjectable({ factory: function ConnectionService_Factory() { return new ConnectionService(); }, token: ConnectionService, providedIn: "root" });
        return ConnectionService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var ConnectionServiceModule = (function () {
        function ConnectionServiceModule() {
        }
        ConnectionServiceModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [],
                        providers: [ConnectionService]
                    },] },
        ];
        return ConnectionServiceModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.ConnectionService = ConnectionService;
    exports.ConnectionServiceModule = ConnectionServiceModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
